### PR TITLE
CASMPET-6321: Add in logic and tests for Multi-Tenancy

### PR DESCRIPTION
**Pull request by ndavidson-hpe**
## Summary and Scope

This adds logic in the keycloak-admin policy to allow/disallow multi tenancy related endpoints.

## Issues and Related PRs

* Resolves [CASMPET-6321](https://jira-pro.its.hpecorp.net:8443/browse/CASMPET-6321)

### Tested on:

  * ashton

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

- Were the install/upgrade-based validation checks/tests run (goss tests/install-validation doc)? y
- Were continuous integration tests run? If not, why? y
- Was upgrade tested? If not, why? y
- Was downgrade tested? If not, why? y
- Were new tests (or test issues/Jiras) created for this change? y

## Risks and Mitigations

No major risks known.

## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [ ] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

### CODE REVIEW

**Review by bklei — APPROVED** [review]

File: kubernetes/cray-opa/tests/opa/keycloak-admin_test.rego.tpl [hunk 1]
```diff
@@ -158,6 +158,25 @@ test_unauth_role {
 
 }
 
+# Multi Tenancy Tests
+
+test_tenat_admin {
```
Comment (jeremy-duckworth) [inline kubernetes/cray-opa/tests/opa/keycloak-admin_test.rego.tpl:163]: typo

File: kubernetes/cray-opa/tests/opa/keycloak-admin_test.rego.tpl [hunk 2]
```diff
@@ -158,6 +158,25 @@ test_unauth_role {
...
+
+  # Check the tenant admin only has access to tenant endpoints
+  allow.http_status == 403 with input as {"attributes": {"request": {"http": {"method": "GET", "path": "/apis/api1", "headers": {"authorization": "Bearer {{ .tenantToken }}", "tenant": "vcluster-coke"}}}}}
```
Comment (jeremy-duckworth) [inline kubernetes/cray-opa/tests/opa/keycloak-admin_test.rego.tpl:166]: A nit, but while we talk about 'coke' vs. 'pepsi' conceptually, suggest we make the mock tenant names generic (not specifying a brand!).

File: kubernetes/cray-opa/templates/policies/keycloak-admin.yaml [hunk 3]
```diff
@@ -111,6 +120,11 @@ data:
         r := parsed_kc_token.payload.resource_access.shasta.roles[_]
     }
 
+    # Get the users realm roles from the JWT token
```
Comment (jeremy-duckworth) [inline kubernetes/cray-opa/templates/policies/keycloak-admin.yaml:123]: Typo, 'user'?

File: kubernetes/cray-opa/templates/policies/keycloak-admin.yaml [hunk 4]
```diff
@@ -79,6 +79,15 @@ data:
         required_admin_roles[r]
     }
 
+    # This checks the JWT if it has tenant-admin role
+    # also checks if the tenant header matches the group
+    allow {
+        roles_for_tenant[r]
```
Comment (jeremy-duckworth) [inline kubernetes/cray-opa/templates/policies/keycloak-admin.yaml:85]: Looking roles_for_user, roles_for_admin, and roles_for tenant, they all look like duplicate code. Can these be refactored towards DRY to add clarity? 

Separately, while the tenant admin case is the only tenant persona we're tracking, please name symbols appropriately to ensure they track the tenant admin case. This should also help others with intuitive understanding of the policy.

File: kubernetes/cray-opa/tests/opa/run_tests/run_tests.go [hunk 5]
```diff
@@ -180,6 +190,20 @@ func main() {
 	}
 	fmt.Println("user token:", userToken)
 
+	args = createTokenArgs{rrole: "tenant-admin", groups: "vcluster-coke-tenant-admin", issuer: keycloakIssuer, aud: shastaAud, typ: "Bearer"}
```
Comment (jeremy-duckworth) [inline kubernetes/cray-opa/tests/opa/run_tests/run_tests.go:193]: Please again be clear that these are tenant admin tests (e.g., TenantToken vs. TenantAdminToken. Also, please reflect the 'admin' role tests in symbol naming and documentation to be 'infrastructure admin'.

File: kubernetes/cray-opa/tests/opa/keycloak-admin_test.rego.tpl [hunk 6]
```diff
@@ -158,6 +158,25 @@ test_unauth_role {
...
+  not allow.http_status with input as {"attributes": {"request": {"http": {"method": "GET", "path": "/apis/bos/v2", "headers": {"authorization": "Bearer {{ .tenantToken }}", "tenant": "vcluster-coke"}}}}}
+
+  # Check that a tenant has no access to other tenants and has the correct headers
+  allow.http_status == 403 with input as {"attributes": {"request": {"http": {"method": "GET", "path": "/apis/bos/v2", "headers": {"authorization": "Bearer {{ .tenantToken }}", "tenant": "vcluster-pepsi"}}}}}
+  allow.http_status == 403 with input as {"attributes": {"request": {"http": {"method": "GET", "path": "/apis/bos/v2", "headers": {"authorization": "Bearer {{ .tenantToken }}"}}}}}
```
Comment (jeremy-duckworth) [inline kubernetes/cray-opa/tests/opa/keycloak-admin_test.rego.tpl:171]: Please add a test where the tenant header is set, but it's an empty string.

**Review by jeremy-duckworth — CHANGES_REQUESTED** [review]

File: kubernetes/cray-opa/templates/policies/keycloak-admin.yaml (hunk 4 above)
Comment (ndavidson-hpe) [reply kubernetes/cray-opa/templates/policies/keycloak-admin.yaml:85]: Based on how opa parses those separate sections are required. They are similar but not duplicate code, there is no way to combine those sections without major reworking existing multi-tenancy code.

File: kubernetes/cray-opa/templates/policies/keycloak-admin.yaml [hunk 7]
```diff
@@ -79,6 +79,15 @@ data:
...
+        required_admin_roles[r]
+        groups := parsed_kc_token.payload.groups[_]
+        trim_suffix(groups, "-tenant-admin") == http_request.headers["tenant"]
```
[the remark this answers is not in the document]
Comment (ndavidson-hpe) [reply kubernetes/cray-opa/templates/policies/keycloak-admin.yaml:88]: This needs a solution outside of OPA, there is no case insensitivity to header names in OPA. There either needs to be documented knowledge of expected header or a filter to adjust prior to hitting opa.

[the remark this answers is not in the document]
Comment (ndavidson-hpe) [reply kubernetes/cray-opa/templates/policies/keycloak-admin.yaml]: Based on new understanding this will be forced lowercase by envoy so there is no need for tests.

**Review by jeremy-duckworth — CHANGES_REQUESTED** [review]
@bklei @ndavidson-hpe I think we should include input validation on the header value, that constraints the header value in the following ways:

1. Cannot be a zero length string. 
2. Has to conform with a character set that is valid for a tenant name. This should match the TAPMS schema. 

Why? Because I think we should try to do some basic input sanitization for upstream services. 

I think this is a global, if this header is present, validate it conforms to a spec for value. Thoughts?

**Comment by jeremy-duckworth** [comment]
> @bklei @ndavidson-hpe I think we should include input validation on the header value, that constraints the header value in the following ways:
> 
> 1. Cannot be a zero length string.
> 2. Has to conform with a character set that is valid for a tenant name. This should match the TAPMS schema.
> 
> Why? Because I think we should try to do some basic input sanitization for upstream services.
> 
> I think this is a global, if this header is present, validate it conforms to a spec for value. Thoughts?

Added header validation.

**Review by jeremy-duckworth — APPROVED** [review]

**Comment by jeremy-duckworth** [comment]
Working to bump OPA to 0.52.0 due to snyk scan failures.

**Comment by jeremy-duckworth** [comment]
Hold while adding 0.52.0-envoy-rootless container to container-images
